File: tests/test_node.py
"""
Unit tests for Node class
Tests the CNode Python bindings
"""
import pytest
import sys
import os

# Add the build directory to Python path
build_dir = os.path.join(os.path.dirname(__file__), '..', 'src', 'out', 'build', 'x64-debug', 'python')
if os.path.exists(build_dir):
    sys.path.insert(0, build_dir)

import pythewheel


@pytest.fixture
def space():
    """Create a Space for testing"""
    return pythewheel.Space()


@pytest.fixture
def node(space):
    """Create a Node for testing"""
    return pythewheel.Node(space, "TestNode", "Test Description")


@pytest.mark.unit
class TestNode:
    """Test cases for Node class"""

    def test_node_creation(self, space):
        """Test basic node creation"""
        node = pythewheel.Node(space, "Test", "Description")
        assert node is not None
        assert node.get_name() == "Test"
        assert node.get_description() == "Description"

    def test_node_without_space(self):
        """Test creating a node without a space"""
        node = pythewheel.Node()
        assert node is not None

    def test_get_set_name(self, node):
        """Test getting and setting node name"""
        node.set_name("NewName")
        assert node.get_name() == "NewName"

    def test_get_set_description(self, node):
        """Test getting and setting node description"""
        node.set_description("New Description")
        assert node.get_description() == "New Description"

    def test_get_set_class(self, node):
        """Test getting and setting node class"""
        node.set_class("TestClass")
        assert node.get_class() == "TestClass"

    def test_initial_activation(self, node):
        """Test initial activation values.
        CNode initializes with m_primaryActivation=0.005 and
        m_secondaryActivation=0.005, so total is 0.01."""
        assert abs(node.get_activation() - 0.01) < 1e-6
        assert abs(node.get_primary_activation() - 0.005) < 1e-6
        assert abs(node.get_secondary_activation() - 0.005) < 1e-6

    def test_set_activation(self, node):
        """Test setting activation value.
        SetActivation splits the delta equally between primary and secondary
        when called without an activator node."""
        node.set_activation(0.5)
        assert abs(node.get_activation() - 0.5) < 1e-6

    def test_activation_splits_primary_secondary(self, node):
        """Test that SetActivation splits delta between primary and secondary.
        Starting from 0.005/0.005, setting total to 0.5 adds delta=0.49.
        Primary gets +0.245, secondary gets +0.245."""
        node.set_activation(0.5)
        assert abs(node.get_primary_activation() - 0.25) < 1e-4
        assert abs(node.get_secondary_activation() - 0.25) < 1e-4

    def test_position(self, node):
        """Test getting and setting position"""
        pos = pythewheel.Vector3D(1.0, 2.0, 3.0)
        node.set_position(pos)
        retrieved_pos = node.get_position()
        assert retrieved_pos[0] == 1.0
        assert retrieved_pos[1] == 2.0
        assert retrieved_pos[2] == 3.0

    def test_link_count_initial(self, node):
        """Test initial link count is zero"""
        assert node.get_link_count() == 0

    def test_link_to_another_node(self, space):
        """Test creating a link between two nodes"""
        node1 = pythewheel.Node(space, "Node1", "First")
        node2 = pythewheel.Node(space, "Node2", "Second")

        node1.link_to(node2, 0.5)

        # Check link was created
        assert node1.get_link_count() >= 1

        # Check we can retrieve the link
        link = node1.get_link_to(node2)
        assert link is not None
        assert abs(link.get_weight() - 0.5) < 1e-6

    def test_get_link_weight(self, space):
        """Test getting link weight between nodes"""
        node1 = pythewheel.Node(space, "Node1", "First")
        node2 = pythewheel.Node(space, "Node2", "Second")

        node1.link_to(node2, 0.75)
        weight = node1.get_link_weight(node2)

        assert abs(weight - 0.75) < 1e-6

    def test_reciprocal_link(self, space):
        """Test that reciprocal links are created by default"""
        node1 = pythewheel.Node(space, "Node1", "First")
        node2 = pythewheel.Node(space, "Node2", "Second")

        node1.link_to(node2, 0.5, True)

        # Both nodes should have links
        assert node1.get_link_count() >= 1
        assert node2.get_link_count() >= 1

    def test_unlink_nodes(self, space):
        """Test unlinking nodes"""
        node1 = pythewheel.Node(space, "Node1", "First")
        node2 = pythewheel.Node(space, "Node2", "Second")

        node1.link_to(node2, 0.5)
        initial_count = node1.get_link_count()

        node1.unlink(node2)

        # Link count should decrease
        assert node1.get_link_count() < initial_count

    def test_parent_child_relationship(self, space):
        """Test parent-child relationships"""
        parent = pythewheel.Node(space, "Parent", "Parent Node")
        child = pythewheel.Node(space, "Child", "Child Node")

        child.set_parent(parent)

        # Check parent-child relationship
        assert child.get_parent() == parent
        assert parent.get_child_count() == 1

    def test_radius(self, node):
        """Test radius calculation based on activation"""
        node.set_activation(0.25)
        radius = node.get_radius()
        # radius = sqrt(activation)
        assert abs(radius - 0.5) < 1e-4


@pytest.mark.integration
class TestNodeActivation:
    """Integration tests for node activation spreading"""

    def test_multiple_nodes_activation(self, space):
        """Test activation with multiple interconnected nodes"""
        # Create a small network
        node1 = pythewheel.Node(space, "A", "Node A")
        node2 = pythewheel.Node(space, "B", "Node B")
        node3 = pythewheel.Node(space, "C", "Node C")

        # Link them
        node1.link_to(node2, 0.5)
        node2.link_to(node3, 0.5)

        # Activate first node
        node1.set_activation(1.0)

        assert abs(node1.get_activation() - 1.0) < 1e-6

File: src/theWheelModelTests/test_node.cpp
// test_node.cpp - Unit tests for CNode class

#include "stdafx.h"
#include <gtest/gtest.h>
#include "Node.h"
#include "Space.h"
#include <cmath>

// ===== Construction =====

TEST(Node, DefaultConstruction)
{
    CNode node;
    EXPECT_EQ(node.GetSpace(), nullptr);
    EXPECT_EQ(node.GetParent(), nullptr);
}

TEST(Node, ConstructionWithName)
{
    CNode node(nullptr, _T("TestNode"));
    CString name = node.GetName();
    EXPECT_STREQ((LPCTSTR)name, _T("TestNode"));
}

TEST(Node, ConstructionWithNameAndDescription)
{
    CNode node(nullptr, _T("TestNode"), _T("A description"));
    CString desc = node.GetDescription();
    EXPECT_STREQ((LPCTSTR)desc, _T("A description"));
}

TEST(Node, InitialActivation)
{
    // Node.cpp:57-58: primary=0.005, secondary=0.005
    CNode node;
    EXPECT_NEAR(node.GetActivation(), 0.01, 1e-6);
    EXPECT_NEAR(node.GetPrimaryActivation(), 0.005, 1e-6);
    EXPECT_NEAR(node.GetSecondaryActivation(), 0.005, 1e-6);
}

// ===== Attributes =====

TEST(Node, SetGetName)
{
    CNode node;
    node.SetName(_T("NewName"));
    EXPECT_STREQ((LPCTSTR)node.GetName(), _T("NewName"));
}

TEST(Node, SetGetDescription)
{
    CNode node;
    node.SetDescription(_T("NewDesc"));
    EXPECT_STREQ((LPCTSTR)node.GetDescription(), _T("NewDesc"));
}

TEST(Node, SetGetClass)
{
    CNode node;
    node.SetClass(_T("MyClass"));
    EXPECT_STREQ((LPCTSTR)node.GetClass(), _T("MyClass"));
}

// ===== Activation =====

TEST(Node, SetActivationPrimarySecondarySplit)
{
    // With pActivator=NULL, PRIM_FRAC=0.5 applies
    CNode node;
    node.SetActivation(1.0f);
    // delta = 1.0 - 0.01 = 0.99
    // primary += 0.5 * 0.99 = 0.005 + 0.495 = 0.5
    // secondary += 0.5 * 0.99 = 0.005 + 0.495 = 0.5
    EXPECT_NEAR(node.GetPrimaryActivation(), 0.5, 1e-4);
    EXPECT_NEAR(node.GetSecondaryActivation(), 0.5, 1e-4);
    EXPECT_NEAR(node.GetActivation(), 1.0, 1e-4);
}

TEST(Node, SetActivationMultipleTimes)
{
    CNode node;
    node.SetActivation(0.5f);
    EXPECT_NEAR(node.GetActivation(), 0.5, 1e-4);
    node.SetActivation(0.8f);
    EXPECT_NEAR(node.GetActivation(), 0.8, 1e-4);
}

// ===== Linking =====

TEST(Node, LinkToCreatesLink)
{
    CNode a, b;
    a.LinkTo(&b, 0.5f, FALSE);
    EXPECT_EQ(a.GetLinkCount(), 1);
    EXPECT_EQ(a.GetLinkAt(0)->GetTarget(), &b);
}

TEST(Node, LinkToReciprocal)
{
    CNode a, b;
    a.LinkTo(&b, 0.5f, TRUE);
    EXPECT_EQ(a.GetLinkCount(), 1);
    EXPECT_EQ(b.GetLinkCount(), 1);
    EXPECT_EQ(a.GetLinkTo(&b)->GetTarget(), &b);
    EXPECT_EQ(b.GetLinkTo(&a)->GetTarget(), &a);
}

TEST(Node, SelfLinkIsIgnored)
{
    // Node.cpp:415-418: self-link returns early
    CNode a;
    a.LinkTo(&a, 0.5f);
    EXPECT_EQ(a.GetLinkCount(), 0);
}

TEST(Node, GetLinkWeight)
{
    CNode a, b;
    a.LinkTo(&b, 0.7f, FALSE);
    EXPECT_FLOAT_EQ(a.GetLinkWeight(&b), 0.7f);
}

TEST(Node, GetLinkToNonExistent)
{
    CNode a, b;
    EXPECT_EQ(a.GetLinkTo(&b), nullptr);
}

TEST(Node, GetLinkWeightNonExistent)
{
    CNode a, b;
    EXPECT_FLOAT_EQ(a.GetLinkWeight(&b), 0.0f);
}

TEST(Node, UnlinkRemovesLink)
{
    CNode a, b;
    a.LinkTo(&b, 0.5f, FALSE);
    EXPECT_EQ(a.GetLinkCount(), 1);
    a.Unlink(&b, FALSE);
    EXPECT_EQ(a.GetLinkCount(), 0);
    EXPECT_EQ(a.GetLinkTo(&b), nullptr);
}

TEST(Node, UnlinkReciprocal)
{
    CNode a, b;
    a.LinkTo(&b, 0.5f, TRUE);
    a.Unlink(&b, TRUE);
    EXPECT_EQ(a.GetLinkCount(), 0);
    EXPECT_EQ(b.GetLinkCount(), 0);
}

TEST(Node, RemoveAllLinks)
{
    CNode a, b, c;
    a.LinkTo(&b, 0.5f, FALSE);
    a.LinkTo(&c, 0.3f, FALSE);
    EXPECT_EQ(a.GetLinkCount(), 2);
    a.RemoveAllLinks();
    EXPECT_EQ(a.GetLinkCount(), 0);
}

// ===== Hierarchy =====

TEST(Node, ParentChild)
{
    // Parent owns children; parent destructor deletes children
    CNode* parent = new CNode(nullptr, _T("Parent"));
    CNode* child = new CNode(nullptr, _T("Child"));
    child->SetParent(parent);

    EXPECT_EQ(parent->GetChildCount(), 1);
    EXPECT_EQ(parent->GetChildAt(0), child);
    EXPECT_EQ(child->GetParent(), parent);

    // parent destructor will delete child
    delete parent;
}

TEST(Node, GetDescendantCount)
{
    CNode* root = new CNode(nullptr, _T("Root"));
    CNode* c1 = new CNode(nullptr, _T("C1"));
    CNode* c2 = new CNode(nullptr, _T("C2"));
    CNode* c1a = new CNode(nullptr, _T("C1a"));

    c1->SetParent(root);
    c2->SetParent(root);
    c1a->SetParent(c1);

    EXPECT_EQ(root->GetDescendantCount(), 3); // c1, c2, c1a
    EXPECT_EQ(c1->GetDescendantCount(), 1);   // c1a
    EXPECT_EQ(c2->GetDescendantCount(), 0);

    delete root; // cascading delete of c1, c2, c1a
}

TEST(Node, ReparentChild)
{
    CNode* p1 = new CNode(nullptr, _T("P1"));
    CNode* p2 = new CNode(nullptr, _T("P2"));
    CNode* child = new CNode(nullptr, _T("Child"));

    child->SetParent(p1);
    EXPECT_EQ(p1->GetChildCount(), 1);

    child->SetParent(p2);
    EXPECT_EQ(p1->GetChildCount(), 0);
    EXPECT_EQ(p2->GetChildCount(), 1);

    delete p1;
    delete p2; // deletes child
}

// ===== Position =====

TEST(Node, DefaultPositionIsZero)
{
    CNode node;
    const CVectorD<3>& pos = node.GetPosition();
    EXPECT_FLOAT_EQ(pos[0], 0.0f);
    EXPECT_FLOAT_EQ(pos[1], 0.0f);
    EXPECT_FLOAT_EQ(pos[2], 0.0f);
}

TEST(Node, SetGetPosition)
{
    CNode node;
    CVectorD<3> newPos(1.0f, 2.0f, 3.0f);
    node.SetPosition(newPos);
    const CVectorD<3>& pos = node.GetPosition();
    EXPECT_FLOAT_EQ(pos[0], 1.0f);
    EXPECT_FLOAT_EQ(pos[1], 2.0f);
    EXPECT_FLOAT_EQ(pos[2], 3.0f);
}

// ===== Radius =====

TEST(Node, RadiusIsSqrtActivation)
{
    CNode node;
    node.SetActivation(0.25f);
    EXPECT_NEAR(node.GetRadius(), sqrt(0.25), 1e-4);
}

TEST(Node, RadiusForActivation)
{
    CNode node;
    EXPECT_NEAR(node.GetRadiusForActivation(0.16f), sqrt(0.16), 1e-4);
}
